### DISCUSSION TITLE
GoogleService-Info.plist location in Expo Managed workflow

### DISCUSSION THREAD

**Discussion by iboxgithub** [invertase/react-native-firebase#6105] [Q&A] [answered]
Hi everyone, 
so I got one issue while doing the Firebase setup as I am trying to stay in a Managed workflow setup. 
To avoid the error message: 
_No Firebase App '[DEFAULT]' has been created - call firebase.initializeApp()_ 
I followed this (https://docs.expo.dev/guides/setup-native-firebase/#ios) asking to put `GoogleService-Info.plist` in the **root folder**, 
but, to keep the ios example, the official rn-firebase docs (https://rnfirebase.io/#generating-ios-credentials) ask to put `GoogleService-Info.plist` in the **ios folder** 

-> anyone knows what is the recommended way?

**Comment by mikehardy** [comment, marked as answer]
Updated - the answer is in here, so will mark this answer, but the part that is *actually* the answer is below:

https://github.com/invertase/react-native-firebase/discussions/6105#discussioncomment-2288924

**Reply by iboxgithub** [reply]
that's the result when it's in the root folder
<img width="802" alt="image" src="https://user-images.githubusercontent.com/5391993/156461968-d91e19b6-e7a9-4fe4-84bc-0e1eeb5975ae.png">

and for the ios/ folder option, I wanted to have the official feedback before trying out different things and ending up with a workaround, thus this thread

**Reply by mikehardy** [reply]
I paid more attention to your links this time - it appears the value for that key is the relative path to the google services plist file - the most authoritative thing I can think to point you to is the implementation of the file resolution, it's here:

https://github.com/invertase/react-native-firebase/blob/6d6cfeae7400f8207d733ed0775a57d7c7cf5552/packages/app/plugin/src/ios/googleServicesPlist.ts#L31

so, your project root, plus the path you specify in the key. It appears you could use any path so long as it is inside your project. If the plist file is found, you're set

If it's not found, you'll see an error, if you're using the config plugin from this module
https://github.com/invertase/react-native-firebase/blob/6d6cfeae7400f8207d733ed0775a57d7c7cf5552/packages/app/plugin/src/ios/googleServicesPlist.ts#L33-L37

**Reply by iboxgithub** [reply]
hi, thanks a lot for your help. 

1. _[Expo method]_ I guess what I am failing to understand is that I configured the app.json with the **GoogleService-Info.plist** relative location (same level), so I assumed it should find the file :(
<img width="414" alt="image" src="https://user-images.githubusercontent.com/5391993/156532612-c326367a-f63b-4d26-a2b0-d53db22c36fe.png">

if it doesn't, then it just seems like the app.json is useless as it doesn't override react-native-firebase default file search in the ios folder no? would this be a bug, or just doc not up to date on the expo side?

2. _[rn-firebase method]_ I also tried to move **GoogleService-Info.plist** in the ios/ folder, updated **AppDelegate.m** with 
`#import <Firebase.h>` and
`[FIRApp configure];`
then ran `npx pod-install`

my `expo run:ios` still fails with the same error message as in my original message :(

**Reply by mikehardy** [reply]
I can't vouch for the Expo docs at all, we have no relationship with them, do not work on their docs and do not consult them ever. I wasn't even aware they have react-native-firebase docs.

What we have here is a couple of config plugins for people that want to use our module with Expo, with open code so you can see how it's working and even alter the code to add console.logs and see what parameters / return values are as it's working. 

If you want to use this module, I would rely on the info in this module. https://rnfirebase.io/#expo

**Reply by iboxgithub** [reply]
thank you, using this: https://rnfirebase.io/#installation-1 was the solution, the command expo prebuild --clean was what was missing. 

for their docs I will get in touch with them once my setup is finished

**I am almost there!**

I still get the error message below though, any idea why the setup didn't fix that, or how to debug it? 
Also, I noticed that AppDelegate.m has been reset after the clean, do I need to modify it?
<img width="775" alt="image" src="https://user-images.githubusercontent.com/5391993/156574507-79239a11-0f69-4fea-a46c-dcc5325d2cd3.png">

**Reply by mikehardy** [reply]
I am unsure, I don't use Expo myself. If that still happens with a completely clean / throwaway attempt / minimal reproduction following instructions, then perhaps there is an error in the documentation, or the plugin, or perhaps the Expo config plugins infrastructure has changed in a way that the docs haven't kept pace with. The plugin is attempting to put a `[FIRApp configure]` line in your AppDelegate.m file though. If that doesn't succeed / if AppDelegate.m doesn't have it, thhen something's wrong. Perhaps @barthap would know?

**Reply by barthap** [reply]
- The docs on `docs.expo.dev` are outdated and they were purposed for `expo-firebase-core` (that is going to be deprecated anyway in favor of RNFB) rather than `react-native-firebase`. They're more applicable if you're using bare workflow. But it would be good to finally update them, PRs are welcome.
- If you're staying in Managed workflow, then running `expo prebuild` should do all modifications in `AppDelegate` for you, as long as you have `plugins` section and Google services path configured in `app.json`. And you shouldn't modify this file manually. The `--clean` flag deletes the `ios` directory and generates it from scratch, while the absence of that flag tries to apply changes incrementally.

**Reply by iboxgithub** [reply]
Hi @barthap thanks for the message, and glad to hear the docs are outdated for this point, I will try to send a PR. 

Nevertheless, if I followed, it should work as my `app.json` has the Google services path configured (see above) -> what do you mean by `plugins` though?
Because as of now, when I run expo prebuild, my `AppDelegate` modifications are removed, the below disappear, and I have the error (No Firebase app) message posted above:
- #import <Firebase.h>
- [FIRApp configure]; in didFinishLaunchingWithOptions function

my app is "clean", there is nothing special in it, I am just trying a regular Firebase file upload

**Reply by barthap** [reply]
In `app.json` you should have the [`plugins`](https://docs.expo.io/versions/latest/config/app/#plugins) section - something like this:
```json
{
  "plugins": ["@react-native-firebase/app", ...]
}
```
it is is documented in [RNFB docs](https://rnfirebase.io/#installation-1) - see the example JSON there. Also, there are some [links to expo docs](https://docs.expo.io/guides/config-plugins/) describing the idea behind config plugins.

**Reply by iboxgithub** [reply]
I missed that, my bad. It works now thanks!
How can I put this question as Answered?

**Reply by mikehardy** [reply]
I marked it as answered with a link up top to the part that is actually the answer (I think), so it's ideally quick+useful for folks. Thanks again @barthap !

**Reply by sarveshhh-m** [reply]
I do have all the setup that mentioned above. but the firebase isn't working.

**Comment by danshilm** [comment]
If you're using the managed workflow, there's no need to run `expo prebuild` (like the Expo docs and react-native-firebase docs mention) unless you're planning on building locally (with `expo run`). If you use EAS build, then you can keep your app in the (fully) managed workflow and use the plugins from `react-native-firebase`, only drawback being that you'd have to use a custom development client instead of the Expo Go app (with `expo start --dev-client`).

And in the (fully) managed workflow, you can keep your `google-services.json` and the `plist` equivalent in the root of the project and point to those files in your `app.json` file with the `android`/`ios`.`config.googleServicesFile` properties. Expo will then handle making sure those are in their proper locations.

**Reply by barthap** [reply]
Everything you stated is true, IMO it's also good to gitignore android/ios directories.

But where do docs say that you have to prebuild?

**Reply by Nantris** [reply]
@danshilm does that mean we can also nest the services files in a folder like `./googleServices/google-services.json`? Firebase says to put it in the root, but for a number of reasons, I'd rather not.

**Reply by mikehardy** [reply]
@Slapbox I think you might just try it and see? If the build does not find the file, it will be immediately obvious as nothing firebase-related will work

**Reply by danshilm** [reply]
@barthap Unfortunately, I can't find the page where I read that anymore; it's been a long time. I just know that the section on the installation section on the rnfirebase docs after the example snippet of the `app.json` file where config plugins are used was from the Expo docs; this sentence specifically
> you need to use the expo prebuild --clean command as described in the ["Adding custom native code"](https://docs.expo.io/workflow/customizing/) guide to rebuild your app with the plugin changes.

Reading that, I started going down the road to "eject" my Expo project, but after messing with expo prebuild and config plugins not working as I expected after reading the Expo docs, it finally clicked. And I presume what I understood from the docs at that time happened to a lot of other users too unfortunately. It's good that the rnfirebase docs now explicitly mention that there's no need to use expo prebuild and go down the "bare workflow", and it also seems that what I saw on the Expo docs isn't there anymore.

@Slapbox yep you can

**Reply by barthap** [reply]
From what I know, the Expo docs are currently being heavily reworked, so my links may be no longer accurate. I'll update them when the docs refactor is mostly finished.

Now you can find most information on these pages:
- [Prebuilding](https://docs.expo.dev/workflow/prebuild/)
- [Expo workflows](https://docs.expo.dev/introduction/managed-vs-bare/)
- [Development builds/clients](https://docs.expo.dev/development/introduction/)

**Reply by sarveshhh-m** [reply]
> googleServicesFile

for iOS, it has to be at **ios.config. googleServicesFile** or **ios. googleServicesFile** ?? I am trying to integrate crashlytics but it is not working.
